triplea-game/triplea
# Replace GameEnginePropertyReader with ProductConfiguration

## Overview

Follow up to #4029.

It was noted that "game engine properties" doesn't really make sense any longer for the following reasons:

1. It only contains a single version property.
1. That version is applicable to more than just the game engine; it applies to every artifact produced from this repo (e.g. game client, lobby server, HTTP server, etc.).

Therefore, this PR generalizes the "game engine version" into a "product version."  Overall, not much changes functionally from #4029.  It's mostly an exercise in renaming things.  Although, there were some build system changes described under Functional Changes below.

## Functional Changes

* Changed how the product version is generated during the Gradle build:
    * _game-core/src/main/resources/META-INF/triplea/product.properties_ is now the Single Source of Truth for the first four components (major, minor, point, micro) of the product version (previously, one had to change both _game_engine.properties_ and _.travis.yml_ when any of these components were updated).
    * _.travis.yml_ is now the Single Source of Truth for the fifth component (build number a.k.a. qualifier) of the product version.
    * Gradle is now the Single Source of Truth for the complete five-component product version (i.e. it merges the value from _product.properties_ and the build number specified in the project property `productVersionQualifier`).  Previously, this was spread around the Gradle buildscripts and _.travis.yml_.

## Refactoring Changes

* Renamed `GameEnginePropertyReader` to `ProductConfiguration` and moved it to the `org.triplea.common.config.product` package.  `org.triplea.common` was chosen because this applies to more than just the game client.  Even though the lobby server has a legitimate dependency on `game-core` at this time, I suspect we'll see more code that fits into a non-game package that applies to multiple projects.  At some point in the future, we'll probably even extract a `common` subproject in the Gradle build.
* Moved the existing types in `g.s.engine.config` to `o.t.common.config` for the same reason as the previous point.
* Applied `version = getProductVersion()` to _all_ Gradle subprojects instead of duplicating it in each subproject buildscript.

## Manual Testing Performed

* Built a full release in my fork and verified the following:
    * All artifacts have the correct version in their file names.
    * All artifacts have the correct version in their fat jar names.
    * Running the headed client from the following artifacts displayed the correct product/engine version on the main menu:
        * Portable build
        * Unix installer
        * Windows x86_64 installer
* Verified running the headed client from Eclipse did not produce any unintended side effects with a full version string of `1.9.0.0.@qualifier@` (see below).

## Additional Review Notes

* Some notes on how the version qualifier is replaced:
    * When you run the client from Eclipse, the engine version will appear as `1.9.0.0.@qualifier@` because the JDT doesn't run the Gradle `processResources` task as part of its own build.  I know IntelliJ has better Gradle integration, so it might not be the same there (if IntelliJ does run `processResources`, you will see `1.9.0.0.dev` instead).
    * When you run the Gradle build locally but do not set the `productVersionQualifier` project property on the command line, the resulting version will be `1.9.0.0.dev`.
    * When you run the Gradle build locally and set the `productVersionQualifier` project property on the command line, the resulting version will be `1.9.0.0.<whatever you specified>`.
* I chose not to rename `ClientContext#engineVersion()` at this time.  Renaming that method alone makes some of the expressions it's used in not read as well because you see things like "if product version != engine version".  Not sure if that could be confusing.  We probably _should_ rename it to `productVersion()`, but I think it would be better to do that in a separate PR where we rename _everything_ that contains the phrase "engine version" to "product version."  However, I'm open to different opinions on this.
